lib/puppet/util/windows/sid.rb

```diff
@@ -131,6 +157,110 @@ def valid_sid?(string_sid)
     end
     module_function :valid_sid?
 
+    ERROR_INSUFFICIENT_BUFFER = 122
+
+    def lookup_account_name(system_name = '.', account_name)
+      # TODO: if system_name is '.', this API doesn't support that and we must use hostname instead
+      FFI::MemoryPointer.from_string_to_wide_string(system_name) do |system_name_ptr|
+        FFI::MemoryPointer.from_string_to_wide_string(account_name) do |account_name_ptr|
+          FFI::MemoryPointer.new(:byte, MAXIMUM_SID_BYTE_LENGTH) do |sid_ptr|
+            FFI::MemoryPointer.new(:dword, 1) do |sid_length_ptr|
+              FFI::MemoryPointer.new(:dword, 1) do |domain_length_ptr|
+                FFI::MemoryPointer.new(:uint32, 1) do |name_use_enum_ptr|
+
+                sid_length_ptr.write_dword(MAXIMUM_SID_BYTE_LENGTH)
+                success = LookupAccountNameW(system_name_ptr, account_name_ptr, sid_ptr, sid_length_ptr,
+                  FFI::Pointer::NULL, domain_length_ptr, name_use_enum_ptr)
+                last_error = FFI.errno
+
+                if (success == FFI::WIN32_FALSE && last_error != ERROR_INSUFFICIENT_BUFFER)
+                  raise Puppet::Util::Windows::Error.new('Failed to call LookupAccountNameW', last_error)
+                end
+
+                FFI::MemoryPointer.new(:lpwstr, domain_length_ptr.read_dword) do |domain_ptr|
+                  if LookupAccountNameW(system_name_ptr, account_name_ptr,
+                      sid_ptr, sid_length_ptr,
+                      domain_ptr, domain_length_ptr, name_use_enum_ptr) == FFI::WIN32_FALSE
+                   raise Puppet::Util::Windows::Error.new('Failed to call LookupAccountNameW')
+                  end
+
+                  domain = domain_ptr.read_wide_string(domain_length_ptr.read_dword)
+
+                  return { :system => system_name,
+                     :account => account_name,
+                     :sid_bytes => sid_ptr.read_bytes(sid_length_ptr.read_dword),
+                     :sid => sid_ptr_to_string(sid_ptr),
+                     :domain_account => domain && !domain.empty? ? "#{domain}\\#{account_name}" : account_name,
```

Comment:
Shouldn't we also include `:domain` here with a value of `.`? Otherwise the two returns do not match.